core/state/state.go

```diff
@@ -0,0 +1,788 @@
+// TODO(weiihann): deal with flush atomicity
+func (s *State) Update(
+	blockNum uint64,
+	update *core.StateUpdate,
+	declaredClasses map[felt.Felt]core.Class,
+) error {
+	if err := s.verifyComm(update.OldRoot); err != nil {
+		return err
+	}
+
+	dirtyClasses := make(map[felt.Felt]core.Class)
+
+	// Register the declared classes
+	for hash, class := range declaredClasses {
+		hasClass, err := HasClass(s.db.disk, &hash)
+		if err != nil {
+			return err
+		}
+
+		if !hasClass {
+			dirtyClasses[hash] = class
+		}
+	}
+
+	// Update the class trie
+	if err := s.updateClassTrie(update.StateDiff.DeclaredV1Classes, declaredClasses); err != nil {
+		return err
+	}
+
+	// Register deployed contracts
+	for addr, classHash := range update.StateDiff.DeployedContracts {
+		exists, err := HasContract(s.db.disk, &addr)
+		if err != nil {
+			return err
+		}
+
+		if exists {
+			return ErrContractAlreadyDeployed
+		}
+
+		contract := newContractDeployed(*classHash, blockNum)
+		obj := newStateObject(s, &addr, &contract)
+		s.stateObjects[addr] = obj
+	}
+
+	// Update the contract fields
+	if err := s.updateContracts(blockNum, update.StateDiff); err != nil {
+		return err
+	}
+
+	newComm, stateUpdate, err := s.commit()
+	if err != nil {
+		return err
+	}
+
+	// Check if the new commitment matches the one in state diff
+	if !newComm.Equal(update.NewRoot) {
+		return fmt.Errorf("state commitment mismatch: %v (expected) != %v (actual)", update.NewRoot, &newComm)
+	}
+
+	s.db.stateCache.PushLayer(&newComm, &stateUpdate.prevComm, &diffCache{
+		storageDiffs:      update.StateDiff.StorageDiffs,
+		nonces:            update.StateDiff.Nonces,
+		deployedContracts: update.StateDiff.ReplacedClasses,
+	})
+
+	if err := s.flush(blockNum, &stateUpdate, dirtyClasses, true); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+// Revert a given state update. The block number is the block number of the state update.
+//
+//nolint:gocyclo
+func (s *State) Revert(blockNum uint64, update *core.StateUpdate) error {
+	// Ensure the current root is the same as the new root
+	if err := s.verifyComm(update.NewRoot); err != nil {
+		return err
+	}
+
+	reverseDiff, err := s.GetReverseStateDiff(blockNum, update.StateDiff)
+	if err != nil {
+		return fmt.Errorf("get reverse state diff: %v", err)
+	}
+
+	// Gather the classes to be reverted
+	v0Classes := update.StateDiff.DeclaredV0Classes
+	v1Classes := update.StateDiff.DeclaredV1Classes
+	classHashes := make([]*felt.Felt, 0, len(v0Classes)+len(v1Classes))
+	classHashes = append(classHashes, v0Classes...)
+	for hash := range v1Classes {
+		classHashes = append(classHashes, &hash)
+	}
+
+	// Revert the classes
+	dirtyClasses := make(map[felt.Felt]core.Class)
+	for _, hash := range classHashes {
+		dc, err := s.Class(hash)
+		if err != nil {
+			return err
+		}
+
+		if dc.At != blockNum {
+			continue
+		}
+
+		if dc.Class.Version() == 1 {
```

Comment:
Please use a type assertion here. I want to remove this method from the interface later on, it is not used almost anywhere